src/VisualStudio/Core/Def/Implementation/LanguageClient/LanguageServerClientEventListener.cs

```diff
@@ -13,19 +19,58 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService
     [ExportEventListener(WellKnownEventListeners.Workspace, WorkspaceKind.Host), Shared]
     internal class LanguageServerClientEventListener : IEventListener<object>
     {
+        private readonly LanguageServerClient _languageServerClient;
+        private readonly Lazy<ILanguageClientBroker> _languageClientBroker;
         private readonly TaskCompletionSource<object> _taskCompletionSource;
 
         public Task WorkspaceStarted => _taskCompletionSource.Task;
 
-        public LanguageServerClientEventListener()
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public LanguageServerClientEventListener(LanguageServerClient languageServerClient, Lazy<ILanguageClientBroker> languageClientBroker)
         {
-            _taskCompletionSource = new TaskCompletionSource<object>();
+            this._languageServerClient = languageServerClient;
+            this._languageClientBroker = languageClientBroker;
+            this._taskCompletionSource = new TaskCompletionSource<object>();
         }
 
+        /// <summary>
+        /// LSP clients do not necessarily know which language servers (and when) to activate as they are language agnostic.
+        /// We know we can provide <see cref="LanguageServerClient"/> as soon as the workspace is started, so tell the
+        /// <see cref="ILanguageClientBroker"/> to start loading it.
+        /// </summary>
         public void StartListening(Workspace workspace, object serviceOpt)
         {
             // mark that roslyn solution is added
-            _taskCompletionSource.SetResult(null);
+            this._taskCompletionSource.SetResult(null);
+
+            Task.Run(() =>
```

Comment:
I think this potentially introduces a race condition. The assumption of

```
this._languageClientBroker.Value.LoadAsync(new LanguageClientMetadata(new string[] { ContentTypeNames.CSharpContentType, ContentTypeNames.VisualBasicContentType }), this._languageServerClient).Forget();
```

is that the registration of the initializing client is happening before the thread yields, ensuring that we have record of your initializing client on focus of the Ctrl+Q box when it's focused so we can await your initialization.

Doing this in a task, there's a small chance that registration won't happen by the time the user focuses the box.

Reply:
So I see that since we're running the task, it won't synchronously complete up to the point where the task gets added to the lsp client collection.

But couldn't that be an issue with continue with?  If the original task has the continuation option to run asynchronously - https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcontinuationoptions?view=netframework-4.8 ( RunContinuationsAsynchronously ).

Reply:
>But couldn't that be an issue with continue with? 

It looks like you removed the ContinueWith?

Reply:
I removed it, what I was saying is that I think it could have been an issue before - since continuewith idt necessarily runs the continuation synchronously.  

This would be fixed if I do Task.RunSynchronously right?  It would run the task synchronously, but wouldn't block since we're calling .Forget() on it.

Reply:
> since continuewith idt necessarily runs the continuation synchronously.

The important thing is that the registration happens before the task tracking solution load completes.

My understanding of .NET async code is that calling an async method without awaiting the task via `Forget()` should cause the method to execute up to the first yield/await.

This should have the desired effect, even in an async code path, so long as the solution load task awaits completion of the calling async code path.

The overarching goal is to make sure your ILanguageClient is registered as loading with the LSP client before the threaded wait dialog disappears.

Reply:
Even though it'd be equivalent, I think it'd be clearer to the code reviewer if we updated the LSP client with a 
```csharp
void BeginLoad()
```
method that does the equivalent through a synchronous signature.

Reply:
per this https://devblogs.microsoft.com/pfxteam/when-executesynchronously-doesnt-execute-synchronously/ I was under the impression that it will just queue the continuation task, which means it would be the same as task.run right?

Reply:
>per this https://devblogs.microsoft.com/pfxteam/when-executesynchronously-doesnt-execute-synchronously/ I was under the impression that it will just queue the continuation task, which means it would be the same as task.run right?

I think that's correct for `ContinueWith`. For `LoadAsync().Forget()`, I would expect up to the first await/yield to happen on the calling thread.

Ultimately, it probably doesn't matter too much since in its current state, Roslyn doesn't wait for projects to load before marking the search complete, so there are scenarios in which the user won't get results early in the session.

Ideally, we'd 1) send what's available 2) when a project finishes loading, send an update with that new batch, but that's a larger change.

Reply:
Discussed offline with @sharwell (thanks!!) - since the workspace started task has already been completed, we can just call LoadAsync.Forget without any continuewith / task.run.  Which should then execute synchronously until the task gets loaded into the collection inside the LSP client.

Reply:
Not sure what you're asking, should we chat in person?